core/src/main/scala/org/apache/spark/shuffle/ShuffleBlockPusher.scala

```diff
@@ -142,7 +142,7 @@ private[spark] class ShuffleBlockPusher(conf: SparkConf) extends Logging {
    * VisibleForTesting
    */
   protected def submitTask(task: Runnable): Unit = {
-    if (BLOCK_PUSHER_POOL != null) {
+    if (BLOCK_PUSHER_POOL != null && !BLOCK_PUSHER_POOL.isShutdown) {
```

Comment:
I think in addition we should catch all kinds exception from the pusher threads and not bubble it up.  I am suggesting to change below to catch all throwables as well. 

 ```
private[shuffle] def tryPushUpToMax(): Unit = {
    try {
      pushUpToMax()
    } catch {
      case e: FileNotFoundException =>
        logWarning("The shuffle files got deleted when this shuffle-block-push-thread " +
          "was reading from them which could happen when the job finishes and the driver " +
          "instructs the executor to cleanup the shuffle. In this case, push of the blocks " +
          "belonging to this shuffle will stop.", e)
    }
  }
```

Reply:
The `tryPushUpToMax` may not be able to  catch the `RejectedExecutionException`, right ?   @otterc  Because,  the `tryPushUpToMax` might  be not called if the BLOCK_PUSHER_POOL has been  shutted down.

Reply:
@weixiuli Yes. I suggested an additional change after looking at this code again. Your fix is good.

Reply:
@otterc  How about catching other exceptions like this?
```
 private[shuffle] def tryPushUpToMax(): Unit = {
    try {
      pushUpToMax()
    } catch {
      case i: InterruptedException =>
        Thread.currentThread().interrupt()
        throw i
      case e: FileNotFoundException =>
        logWarning("The shuffle files got deleted when this shuffle-block-push-thread " +
          "was reading from them which could happen when the job finishes and the driver " +
          "instructs the executor to cleanup the shuffle. In this case, push of the blocks " +
          "belonging to this shuffle will stop.", e)
      case other => throw other
    }
  }
```

Reply:
So, the reason I am thinking of not bubbling up any type of exceptions from pusher threads is because:
1. push is best-effort so if there are any exceptions there it should not interfere with the rest of the executor tasks.
2. the exceptions in these pusher threads can  also shutdown executor and create confusion.
IMO, I think we should just catch and log all the `Throwables` here: 
```
 private[shuffle] def tryPushUpToMax(): Unit = {
    try {
      pushUpToMax()
    } catch {
      case t: Throwable =>
        logWarning("There was .... Stopping block push", e)
    }
  }
```

Reply:
Instead of `Throwable`, let us limit to `NonFatal`

Reply:
> Instead of `Throwable`, let us limit to `NonFatal`
should we cach the exception like that ? @mridulm @otterc 
```
  private[shuffle] def tryPushUpToMax(): Unit = {
    try {
      pushUpToMax()
    } catch {
      case NonFatal(e) =>
        logWarning("Failed to tryPushUpToMax", e)
    }
  }
```